### DISCUSSION TITLE
Can I use WindowsAppSDK in Winforms?

### DISCUSSION THREAD

**Discussion by MyTechApplications** [microsoft/WindowsAppSDK#2404] [Q&A]
Can I use WindowsAppSDK in Winforms?

**Comment by castorix** [comment]
Yes, but only with .NET Core (5, 6, ...) atm (while there is no XAML Island for WinUI 3)
(by using _Microsoft.UI.Xaml.Application.Start_ and creating the Windows Form inside)

**Reply by MyTechApplications** [reply]
can you please tell me how?

**Reply by ghost1372** [reply]
explore samples
https://github.com/microsoft/WindowsAppSDK-Samples

**Reply by MyTechApplications** [reply]
I did and now I am having a problem in [winforms_unpackagedapp ](https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/Windowing/cs-winforms-unpackaged):
Exception thrown: 'System.Runtime.InteropServices.COMException' in Microsoft.WindowsAppRuntime.Bootstrap.Net.dll

**Reply by DrusTheAxe** [reply]
What version of WindowsAppSDK are you building with?

What version(s) of WinAppSDK are installed? You can enumerate them via `powershell -c $(get-appxpackage micro*win*apprun*).packagefullname`

Are you running under a debugger? If so the debugger's output window/pane should show error(s).

Have you run WinAppSDK runtime's installer? Just hitting F5 won't suffice (VS doesn't know about how to install the WinAppSDK runtime for unpackaged apps).

WinAppSDK 1.1-preview1 added some error information to help diagnose bootstrap errors, and the upcoming 1.1-preview2 adds significantly more.

**Comment by michael-hawker** [comment]
So since WinUI 3 is part of the WindowsAppSDK, do those dlls just get excluded by the SDK build process when part of a WinForms or WPF app?

If other library authors want to build on top of the WindowsAppSDK for these non-WinUI 3 scenarios, is there anything special they need to worry about for these scenarios?

**Reply by DrusTheAxe** [reply]
No problemo.

Windows App SDK doesn't support any sort of 'subsetting' today. When you use WinAppSDK you have all of it available. If you don't use WinUI3 it's still present and available, same for MRTCore, AppLifecycle, PowerStateNotifications and the rest of the functionality.

What you choose to use is up to you. Having the functionality present doesn't require an app to do anything special not to use it. Use WinForms, Qt, MFC, WxWidgets, printf() 🙂 or whatever other libraries as you desire.

**Reply by michael-hawker** [reply]
Thanks for the quick response @DrusTheAxe! I guess my question is more around the binary size of the resulting app.

So, if a developer creates a WinForms or WPF app but doesn't use any WinUI 3 APIs, do the WinUI 3 dlls still get copied over to the output of the app? Or are they automatically trimmed?

Similarly, if we have a class library that depends on the WindowsAppSDK, but doesn't call any WinUI 3 APIs and gets included in that app, would that same trimming (if any) occur still?

If the library does call a WinUI 3 APIs, but the app doesn't use that API, I'd assume the trimming would probably not happen in that case or maybe only on newer .NET runtimes with better trimming support?

**Reply by DrusTheAxe** [reply]
>if a developer creates a WinForms or WPF app but doesn't use any WinUI 3 APIs, do the WinUI 3 dlls still get copied over to the output of the app?

That depends

* If you use WinAppSDK's runtime deployed as MSIX packages then NO 
* If you use WinAppSDK's runtime deployed via the 'Self-Contained' option then YES

A handful of WinAppSDK artifacts are (always) redistributable e.g. the Bootstrapper DLL and WinAppSDK's C# projection DLLs (due to the way .NET and C#/WinRT work).

If you use WinAppSDK via MSIX packages (the default) then you need those packages installed on the system for your app to run, but that's more like the CRT - needs to be installed on the machine but shared by any who need it (and if it's already present you don't need to install anything). No need for each app to have its own copy.

If you use WinAppSDK via the Self-Contained option, then WinAppSDK's runtime is copied to your app/project and must be distributed with your app for your app to run. That is the point after all - a share-nothing private copy. See [Windows App SDK deployment guide for self-contained apps](https://docs.microsoft.com/windows/apps/package-and-deploy/self-contained-deploy/deploy-self-contained-apps) for more details.

This is the classic tradeoff of size and servicing responsibility in exchange for control. As in all great Choose-Your-Own-Adventures you must choose the path you think best suits your needs :-)

>Or are they automatically trimmed?

No. As mentioned, there is no subsetting or 'trimming' today of WinAppSDK. 

If 'trimming' or subsetting (of any kind) interests you please share your feedback via our [productboard portal](https://portal.productboard.com/winappsdk/) for consideration and prioritization. Thanks!

**Comment by Gaoyifei1011** [comment]
> Can I use WindowsAppSDK in Winforms?

Whether it is.NET framework or.NET core, you can use Windows app sdk on winforms, you can refer to the official example under.NET core, and you can refer to the content of my project. GetStoreAppWebView is implemented using Winui3 islands with winform(.net framework 481)

无论是.net framework还是.net core，您都可以使用Windows app sdk在winforms上，.net core下的您可以参照官方示例，.net framework您可以参考我的项目里面的内容，GetStoreAppWebView就是使用的Winui3 islands with winform(.net framework 481) 实现的。

The only downside is that when used under the.NET framework, there is no XAML intellisense, which means that you need to carefully check what you type into the XAML page.
唯一的缺点是，.net framework平台下使用时，没有XAML 智能感知，这意味着您需要仔细检查您在XAML页面输入的内容。

https://github.com/Gaoyifei1011/GetStoreApp
